### PR TITLE
Failover changes

### PR DESCRIPTION
**Pull request by ConalCosgrove**
## Bot Failover Changes
This update will ~~hopefully~~ allow us to utilize redis to load balance responses to websocket events coming from refocus. Currently if two instances of the same bot are connected to refocus then both will respond to each event: eg. they will both page someone resulting in two pages

- ## Cache.js
   - Simple class to abstract interfacing with the redis module

- ## refocus-bdk-server.js
   - Added logic to create cache object if all 3 necessary variables are available in config: 
      - `redisCacheHost`
      - `redisCachePort`
      - `useRedisCache: true`
   - Updated socket.on() calls to check if event has already been handled. We do this by making a single `set` call to redis with the `NX` flag set. This operation creates the key value pair if the key does not exist and returns `OK`. If the key already exists then this is a noop and `null` is returned - if we receive 'null' back from redis then we know we need not respond to this event.

### CODE REVIEW

File: Cache.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    this.client.on('ready', () => {
+      this.logger.info(`Connected to redis at ${host}:${port}`);
+    });
+  }
+
+  /**
+   * @param {string} realtimeEventId - event from websocket stream
+   * @param {string} updatedAt - timestamp, avoids ignoring events with same id
+   * @param {string} type - one of botAction, botData, event
```
Comment (cassiodias) [inline Cache.js:22]: nice one using type

File: Cache.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+   * @returns {Promise}
+   */
+  hasBeenConsumed(realtimeEventId, updatedAt, type) {
+    return new Promise((resolve) => {
+      const key = `${type}${realtimeEventId}${updatedAt}`;
+      this.client.set(key, '1', 'NX', 'ex', timeToLiveInSeconds, (err, res) => {
```
Comment (cassiodias) [inline Cache.js:28]: what is '1', 'NX', 'ex' can we create more meaningful names here?

File: refocus-bdk-server.js [hunk 3]
```diff
@@ -188,46 +204,87 @@ module.exports = (config) => {
...
     const socket = io.connect(connectUrl, opts);
 
-    socket.on(initalizeEventName, (data) => {
+    socket.on(initalizeEventName, async (data) => {
       const eventData = JSON.parse(data);
       const room = eventData[initalizeEventName];
+      const { id, updatedAt } = room;
+      const hasAlreadyBeenConsumed = cache &&
+        await cache.hasBeenConsumed(id, updatedAt, initalizeEventName);
+      if (hasAlreadyBeenConsumed) {
```
Comment (cassiodias) [inline refocus-bdk-server.js:213]: could inline this 'if () return';

File: refocus-bdk-server.js [hunk 4]
```diff
@@ -188,46 +204,87 @@ module.exports = (config) => {
...
+        return;
+      }
       app.emit('refocus.internal.realtime.bot.namespace.initialize', room);
-      log.realtime('New Room', room);
     });
 
-    socket.on(settingsChangedEventName, (data) => {
+    socket.on(settingsChangedEventName, async (data) => {
       const eventData = JSON.parse(data);
       const room = eventData[settingsChangedEventName];
```
Comment (cassiodias) [inline refocus-bdk-server.js:221]: is this a room?

File: refocus-bdk-server.js [hunk 5]
```diff
@@ -188,46 +204,87 @@ module.exports = (config) => {
...
+      const { id, updatedAt } = room;
```
Comment (cassiodias) [inline refocus-bdk-server.js:222]: is this a room ID or event ID?

File: refocus-bdk-server.js [hunk 6]
```diff
@@ -188,46 +204,87 @@ module.exports = (config) => {
...
+        return;
+      }
       app.emit('refocus.bot.actions', action);
     });
 
-    socket.on(botActionsUpdate, (data) => {
+    socket.on(botActionsUpdate, async (data) => {
       const eventData = JSON.parse(data);
       const action = eventData[botActionsUpdate].new;
+      const { id, updatedAt } = action;
```
Comment (cassiodias) [inline refocus-bdk-server.js:246]: 246 to 251 seems to be similar to the rest, can we extract a method?

File: Cache.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
+const redis = require('redis');
+const defaultHost = '127.0.0.1';
+const defaultPort = 6379;
+const timeToLiveInSeconds = 60;
+
+class Cache {
```
Comment (cassiodias) [inline Cache.js:6]: One thing you could do here is to remove the knowledge about the client implementation (Redis or Memcache), Cache could receive the client and you could create a test to validate has been consumed checking the behavior, true, false, null, exception.

The creation of the client can be handled by bdk server line 157 or even creating a handler factory

**Review by cassiodias — CHANGES_REQUESTED** [review]

File: Cache.js (hunk 2 above)
Comment (ConalCosgrove) [reply Cache.js:28]: '1' is just the value that is being set for the key.
'NX' is a flag that indicates that it should only set if the kv pair does not exist (NX). 
'ex' is the param for 'expires' 

Will refactor them into variables

File: refocus-bdk-server.js (hunk 3 above)
Comment (ConalCosgrove) [reply refocus-bdk-server.js:213]: for sure - .eslint doesn't like this but can change that too

File: refocus-bdk-server.js (hunk 5 above)
Comment (ConalCosgrove) [reply refocus-bdk-server.js:222]: good point, it's a roomID - makes me think that if multiple bots use the same redis and are responding to the same event then we'll have trouble. Think I'll need to include bot name in key as well

File: refocus-bdk-server.js (hunk 4 above)
Comment (ConalCosgrove) [reply refocus-bdk-server.js:221]: yeah, this event is fired when the settings of a room are changed eg. it goes from `active` to `inactive`. the event includes the entire room object.

File: refocus-bdk-server.js (hunk 5 above)
Comment (cassiodias) [reply refocus-bdk-server.js:222]: yes, bot name solves the problem, I thought the event would have its own id... something like room.event.id

File: refocus-bdk-server.js (hunk 3 above)
Comment (cassiodias) [reply refocus-bdk-server.js:213]: yep, change .eslint :)

File: refocus-bdk-server.js [hunk 8]
```diff
@@ -188,46 +204,74 @@ module.exports = (config) => {
...
       const action = eventData[botActionsAdd];
+      const { id, updatedAt } = action;
+      const hasAlreadyBeenConsumed = cache &&
+        await cache.hasBeenConsumed(id, updatedAt, botName, botActionsAdd);
+      if (hasAlreadyBeenConsumed) return;
       app.emit('refocus.bot.actions', action);
     });
 
-    socket.on(botActionsUpdate, (data) => {
+    socket.on(botActionsUpdate, async (data) => {
```
Comment (cassiodias) [inline refocus-bdk-server.js:238]: I'd put the same code of those 4 .on in one method that can be called as if (hasAlreadyBeenConsumed(param1, ...)) - they are the same, just the 3rd argument changes

**Review by cassiodias — COMMENTED** [review]
Looks good, just minor comments

**Review by cassiodias — APPROVED** [review]

**Review by LongweiDeng — APPROVED** [review]

**Comment by LongweiDeng** [comment]
Good job on the cache!

**Review by diogo-pessoa — APPROVED** [review]

**Pull request merged**